client/fs_endpoint.go

```diff
@@ -0,0 +1,921 @@
+	go func() {
+		for {
+			if _, err := conn.Read(nil); err != nil {
+				if err == io.EOF {
+					cancel()
+					return
+				}
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	var streamErr error
+OUTER:
+	for {
+		select {
+		case streamErr = <-errCh:
+			break OUTER
+		case frame, ok := <-frames:
+			if !ok {
+				break OUTER
+			}
+
+			var resp cstructs.StreamErrWrapper
+			if req.PlainText {
+				resp.Payload = frame.Data
+			} else {
+				if err = frameCodec.Encode(frame); err != nil {
+					streamErr = err
+					break OUTER
+				}
+
+				resp.Payload = buf.Bytes()
+				buf.Reset()
+			}
+
+			if err := encoder.Encode(resp); err != nil {
+				streamErr = err
+				break OUTER
+			}
+		case <-ctx.Done():
+			break OUTER
+		}
+	}
+
+	if streamErr != nil {
+		f.handleStreamResultError(streamErr, helper.Int64ToPtr(500), encoder)
+		return
+	}
+}
+
+// logs is is used to stream a task's logs.
+func (f *FileSystem) logs(conn io.ReadWriteCloser) {
+	defer metrics.MeasureSince([]string{"client", "file_system", "logs"}, time.Now())
+	defer conn.Close()
+
+	// Decode the arguments
+	var req cstructs.FsLogsRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&req); err != nil {
+		f.handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// Check read permissions
+	if aclObj, err := f.c.ResolveToken(req.QueryOptions.AuthToken); err != nil {
+		f.handleStreamResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil {
+		readfs := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityReadFS)
+		logs := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityReadLogs)
+		if !readfs && !logs {
+			f.handleStreamResultError(structs.ErrPermissionDenied, nil, encoder)
+			return
+		}
+	}
+
+	// Validate the arguments
+	if req.AllocID == "" {
+		f.handleStreamResultError(allocIDNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if req.Task == "" {
+		f.handleStreamResultError(taskNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	switch req.LogType {
+	case "stdout", "stderr":
+	default:
+		f.handleStreamResultError(logTypeNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	switch req.Origin {
+	case "start", "end":
+	case "":
+		req.Origin = "start"
+	default:
+		f.handleStreamResultError(invalidOrigin, helper.Int64ToPtr(400), encoder)
+		return
+	}
+
+	fs, err := f.c.GetAllocFS(req.AllocID)
+	if err != nil {
+		var code *int64
+		if strings.Contains(err.Error(), "unknown allocation") {
+			code = helper.Int64ToPtr(404)
+		} else {
+			code = helper.Int64ToPtr(500)
+		}
+
+		f.handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	alloc, err := f.c.GetClientAlloc(req.AllocID)
+	if err != nil {
+		var code *int64
+		if strings.Contains(err.Error(), "unknown allocation") {
```

Comment:
Doing a string search, while it works now is brittle. Consider returning an Error instance like `ErrAllocNotFound` from GetClientAlloc

Reply:
Added helpers for returning and checking errors for all object types (node, eval, job, alloc, deployment)